docs/source/reference/types.rst

```diff
@@ -366,7 +369,30 @@ Type annotations
       ...         self.x += 1
       ...         return old_val
       ...
-      >>> numba.extending.as_numba_type(Counter)
+      >>> numba.core.typing.as_numba_type(Counter)
       instance.jitclass.Counter#11bad4278<x:int64>
 
+   For cases where Numba requires more type information than what Python type
+   annotations support (e.g. Numpy arrays), or where a different Numba type is
+   desired for the same Python type (e.g. int32 vs int64),
+   ``typing_extensions.Annotated`` can be used.
+   Valid patterns are:
+
+      * ``Annotated[py_type, nb_type]``: ``as_numba_type`` will return ``as_numba_type(nb_type)``
+      * ``Annotated[np.ndarray, dtype, ndim]``: ``as_numba_type`` will return ``Array(dtype=as_numba_type(dtype), ndim=ndim, layout="A")``
+      * ``Annotated[np.ndarray, dtype, ndim, layout]``: ``as_numba_type`` will return ``Array(dtype=as_numba_type(dtype), ndim=ndim, layout=layout)``
```

Comment:
I don't think this is a good idea - why not just require the user to use the normal syntax for array types rather than inventing a new one?

Reply:
With these options, the user doesn't need to use any numba types for some cases.  For instance you can do
```
Annotated[np.ndarray, float, 2]
```
I think this lowers the barrier for entry with new users... they don't even have to read anything about numba types and what `__getitem__` does.  By just using `Annotated[np.ndarray, {float, int, complex, bool}, N]` they can get pretty far.

Reply:
I don't think it's numba's place to advocate for a way to statically type numpy arrays. There are ongoing discussions in numpy about how to annotate arrays with shape and dtype information, inventing a new and different protocol to whatever is standardised there will IMO just cause confusion. `Annotated[np.ndarray, float, 2]` is barely shorter than `Annotated[np.ndarray, numba.Array(float, 2)]`, and the later makes it very clear to the reader that the syntax is numba-specific.

Reply:
I'd also prefer to not invent new ways to annotate ndarrays. `Annotated[np.ndarray, numba.Array(float, 2)]` looks more understandable than `Annotated[np.ndarray, float, 2]`.